### PR TITLE
Added Support for Reading Unfiltered Cell Ranger Feature Barcode Matrices (MEX Format)

### PR DESCRIPTION
**Pull request by Gautam8387**
# Description:
This PR introduces functionality to support the reading and processing of unfiltered Cell Ranger Feature Barcode Matrices (MEX format). Implemented the following:
- Default filtering cutoff at 500
- Implemented `read_header()` and `process_batch()`
- Implemented `_get_valid_barcodes()`
- Assertions for above method
- Implemented `rename_batches()` (to be used in `consume()`)
- Implemented `consume()` as per `_get_valid_barcodes()`

## 1. Header Parsing:
- The `read_header()` method reads the matrix header, extracting essential metadata (`nFeatures`, `nCells`, and `nCounts`) which is necessary for downstream processing.
## 2. Batch Processing
- The `process_batch()` method processes chunks of data in batches, aggregating counts per barcode and filtering out barcodes below a specified cutoff value (default: 500).
## 3. Get Valid Barcode:
- The method uses `pl.scan_csv` to scan the input matrix file (self.matFn), skipping the first three rows (header information) and assuming no header in the data.
- The matrix is divided into chunks based on the `lines_in_mem` parameter, ensuring that only a manageable portion of the data is loaded into memory at a time.
- The total number of chunks is calculated based on the total number of counts (`nCounts`) and the `lines_in_mem` parameter.
- If a chunk’s barcodes exceed the current batch boundary, it processes the accumulated data (`dfs`) and filters out background barcodes using the `process_batch` method.
- The valid barcodes are stored in `valid_idx`, and the method prepares for the next batch.
- Care is taken to handle barcodes that span across multiple chunks. This is managed by checking if the current chunk’s last barcode exceeds the current batch boundary and adjusting accordingly.
- If the chunk is within the batch boundary, it is accumulated in `dfs` for processing in the next iteration or at the end of the loop.
- After iterating through all chunks, any remaining data is processed to ensure all valid barcodes are identified and filtered.
- The method returns a sorted and unique list of valid barcodes after concatenating all valid indices collected across batches.
- An assertion is included to ensure that the total number of processed rows matches the expected count (`nCounts`), ensuring data integrity throughout the operation.
## 4. Rename Batches
Renames barcodes in a batch to a contiguous sequence of indices, starting from 0. This is useful for maintaining consistency when processing large datasets in chunks.
- Concatenates all `polar.DataFrames` in collect into a single `polar.DataFrame`.
- Extracts the `barcode` column and creates a unique list of barcodes and maps each unique barcode to a new index using a dictionary (`count_hash`).
- Replaces the original `barcode` values in the `DataFrame` with these new indices.
- Returns a `NumPy` array of the modified DataFrame with barcodes renamed to sequential indices.

## 5. Consume
Yields sparse matrices in chunks from an MTX file after filtering and processing the data in batches. This function is designed to handle large datasets efficiently.
- Reads the MTX file in batches using Polars, skipping header rows and applying a schema to the columns.
- Filters the data to include only barcodes present in `self.validBarcodeIdx`.
- Accumulates unique barcodes from the current chunk and checks if the accumulated unique barcodes exceed `batch_size`.
- If yes, processes the data, renames the barcodes, converts it to a sparse matrix, and yields it. If no, continues to accumulate chunks.
- Handles the remaining data after the main loop, ensuring all data is processed.
- Yields sparse matrices (`coo_matrix`), each corresponding to a processed batch of data.

# Testing:
- All unit test passed
- Uploaded additional notebook showcasing the correctness on existing and new dataset: [Link](https://github.com/Gautam8387/scarf-pr-exp/tree/main/01-valid-barcode)

# Documentation:
- Added docstrings to all new methods.
- Added comments when necessary
- Additional documentation on [Link](https://github.com/Gautam8387/scarf-pr-exp/tree/main/01-valid-barcode)

# Updates:
- Updated the head reader method to include `comment_prefix` instead of `skip_rows`. Updated other polar readers as well.
- Added test for empty barcodes. Added small dataset for empty barcode test

### CODE REVIEW

File: scarf/readers.py [hunk 1]
```diff
@@ -402,28 +492,76 @@ def to_sparse(self, a: np.ndarray, dtype) -> coo_matrix:
             dtype=dtype,
         )
 
+    def cell_names(self) -> List[str]:
```
Comment (oskbor) [inline scarf/readers.py:495]: is `cell_names` used in other readers? Otherwise I propose the name `barcodes`

File: scarf/readers.py [hunk 2]
```diff
@@ -382,6 +392,86 @@ def _read_dataset(self, key: Optional[str] = None):
...
+        """Returns a list of valid barcodes after filtering out background barcodes.
+
+        Args:
+            filtering_cutoff: The cutoff value for filtering out background barcodes.
+            batch_size: The number of barcodes to process in each batch.
+            lines_in_mem: The number of lines to read into memory
+        """
+        test_counter = 0
+        matrixIO = pl.scan_csv(
+            self.matFn, skip_rows=3, separator=self.sep, has_header=False,
```
Comment (oskbor) [inline scarf/readers.py:431]: .mtx can have several lines of headers (all starting with %), so it could potentially be brittle with a fixed `skip_rows`. 
https://math.nist.gov/MatrixMarket/formats.html

Maybe you can set the `comment_prefix` argument to `%` and test if it will work without a hard coded `skip_rows`?

**Review by oskbor — CHANGES_REQUESTED** [review]
This looks great! 💪
Could we add a test case for a mtx import with empty barcodes?
Just to make sure this code is exercised a bit by the test suite.
I have one suggestion regarding the parsing of the header.

Great job Gautam! 🙌

File: scarf/readers.py [hunk 3]
```diff
@@ -382,6 +392,86 @@ def _read_dataset(self, key: Optional[str] = None):
...
+        start = 1
+        dfs = pl.DataFrame()
+        for i in tqdmbar(
+            range(nChunks), desc="Filtering out background barcodes"
+        ):
+            chunk = matrixIO.slice(i*lines_in_mem, lines_in_mem).collect()
+            # Check if we've reached or exceeded the current batch boundary
+            if (chunk[-1]['barcode'][0] - start) >= batch_size: # If the last "cell id" is greater than the start + batch size
+                # Filter rows in the current chunk that belong to the current batch
+                idx = np.array(chunk['barcode'] < (batch_size + start))  # This is the crucial line. This makes sure that if any cell ID is spread over multiple chunks, it is not missed, as any cell ID that is less than the batch size + start is included.
```
Comment (oskbor) [inline scarf/readers.py:448]: I'm not well versed in Python dataframe syntax;  are we depending on the barcodes being ordered? Not sure that holds true.

File: scarf/readers.py (hunk 1 above)
Comment (oskbor) [reply scarf/readers.py:495]: Seems `cell_names` is used!

File: scarf/readers.py (hunk 2 above)
Comment (Gautam8387) [reply scarf/readers.py:431]: Thanks for pointing this out! I will edit it. 😊
I was following [this](https://kb.10xgenomics.com/hc/en-us/articles/115000794686-How-is-the-MEX-format-used-for-the-gene-barcode-matrices) convention.

File: scarf/readers.py (hunk 3 above)
Comment (Gautam8387) [reply scarf/readers.py:448]: Yes, the barcodes are ordered in ascending order in the `matrix.mtx` file. The order may not be continuous in raw, (for example, it may start from 1 and next entry as 23) but ascending for sure.

Comment (oskbor) [reply scarf/readers.py:459]: Ah right, it’s all indexes in the mtx file

**Review by oskbor — APPROVED** [review]
Looks good to me, check with Parashar if he wants to review before merging

**Review by parashardhapola — APPROVED** [review]
